go/ct/gen/block_context.go

```diff
@@ -13,43 +13,153 @@
 }
 
-func (*BlockContextGenerator) Generate(rnd *rand.Rand, revision common.Revision) (st.BlockContext, error) {
+func (b *BlockContextGenerator) Generate(assignment Assignment, rnd *rand.Rand, revision common.Revision) (st.BlockContext, error) {
 	baseFee := common.RandU256(rnd)
+	blockNumberRangeSolver := NewRangeSolver(uint64(0), math.MaxUint64)
 
 	revisionNumber, err := common.GetForkBlock(revision)
 	if err != nil {
 		return st.BlockContext{}, err
 	}
-	revisionNumberRange, err := common.GetBlockRangeLengthFor(revision)
+
+	blockNumberRangeSolver.AddLowerBoundary(revisionNumber)
+	revisionNumberLength, err := common.GetBlockRangeLengthFor(revision)
 	if err != nil {
 		return st.BlockContext{}, err
 	}
-	var randomOffset uint64
-	if revisionNumberRange != 0 {
-		randomOffset = rnd.Uint64n(revisionNumberRange)
-	} else {
-		randomOffset = rnd.Uint64()
+	var blockNumber uint64
+	if revisionNumberLength != 0 {
+		blockNumberRangeSolver.AddUpperBoundary(revisionNumber + revisionNumberLength)
 	}
-	blockNumber := revisionNumber + randomOffset
 
-	chainId := common.RandU256(rnd)
-	coinbase, err := common.RandAddress(rnd)
+	// if any relevant variable is already bound, then we need to constraint the block number to be generated.
+	for _, offsetConstraint := range b.variablesOffsetConstraints {
+		if assignedValue, isBound := assignment[offsetConstraint.lower.variable]; isBound {
+
+			assignedValueI64 := int64(assignedValue.Uint64())
+			lowerBound := assignedValueI64 - offsetConstraint.upper.offset
+			upperBound := assignedValueI64 + offsetConstraint.lower.offset - 1
+
+			// if the range is only one number ( aka diff 2 ), then the assigned value must be within the range.
+			diff := offsetConstraint.upper.offset - offsetConstraint.lower.offset
+			isFixedValue := diff == 2 || diff == -2
+			isSameAsfixed := offsetConstraint.lower.offset-1 == assignedValueI64
+			if isFixedValue && !isSameAsfixed {
+				return st.BlockContext{}, ErrUnsatisfiable
+			} else if isFixedValue && isSameAsfixed {
+				blockNumberRangeSolver.AddEqualityConstraint(uint64(upperBound - 1))
+			}
+			// in case of desired out of range and bound value too small.
+			if lowerBound < 0 {
+				lowerBound = 0
+			}
+
+			blockNumberRangeSolver.AddLowerBoundary(uint64(lowerBound))
+			blockNumberRangeSolver.AddUpperBoundary(uint64(upperBound))
+
+			// if the range solver is now unsatisfiable, it is most likely because the assigned value is out of the range
+			// of the first number of the current revision - 256.
+			if !blockNumberRangeSolver.IsSatisfiable() {
+				return st.BlockContext{}, ErrUnsatisfiable
+			}
+		}
+	}
+
+	// generate block number
+	blockNumber, err = blockNumberRangeSolver.Generate(rnd)
 	if err != nil {
 		return st.BlockContext{}, err
 	}
+
+	variableRangeSolver := NewRangeSolver(math.MinInt64, int64(blockNumber))
+	baseSolver := NewRangeSolver(math.MinInt64, int64(blockNumber))
```

Comment:
Construction of this solver is trivial, I would construct a new one on each iteration.